### PR TITLE
JAMES-3757 IMAP/SMTP OIDC extensions should support impersonation

### PR DESCRIPTION
**Pull request by vttranlina**
Jira: https://issues.apache.org/jira/browse/JAMES-3757

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/exception/ForbiddenDelegationException.java [hunk 1]
```diff
@@ -19,5 +19,24 @@
...
+        this.delegateUser = delegateUser;
+    }
+
+    public Username givenUser() {
+        return givenUser;
+    }
+
+    public Username delegateUser() {
+        return delegateUser;
+    }
```
Comment (quantranhong1999) [inline mailbox/api/src/main/java/org/apache/james/mailbox/exception/ForbiddenDelegationException.java:41]: Where is this be used? I do not see one yet. Maybe you want to use it for tests in the next commits?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java [hunk 2]
```diff
@@ -86,37 +91,38 @@ protected void doAuth(AuthenticationAttempt authenticationAttempt, ImapSession s
...
+            provisionInbox(session, mailboxManager, mailboxSession);
+        } catch (BadCredentialsException e) {
+            manageFailureCount(session, request, responder);
         } catch (UserDoesNotExistException e) {
-            LOGGER.info("User {} does not exist", otherUser, e);
+            LOGGER.info("User does not exist", e);
             no(request, responder, HumanReadableText.USER_DOES_NOT_EXIST);
         } catch (ForbiddenDelegationException e) {
-            LOGGER.info("User {} is not delegated by {}", givenUser, otherUser, e);
+            LOGGER.info("Delegate forbidden", e);
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java:125]: - This method has been extracted without `givenUser, otherUser` -> can not get it here 
- ForbiddenDelegationException already contain givenUser/otherUser info. 
```java
 public ForbiddenDelegationException(Username givenUser, Username delegateUser) {
        super("User " + givenUser.asString() + " is not delegated by " + delegateUser.asString());
        this.givenUser = givenUser;
        this.delegateUser = delegateUser;
    }
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/exception/ForbiddenDelegationException.java (hunk 1 above)
Comment (vttranlina) [reply mailbox/api/src/main/java/org/apache/james/mailbox/exception/ForbiddenDelegationException.java:41]: IMO, it is a placeholder, We can easy get it when needed (maybe in feature)

File: protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java [hunk 3]
```diff
@@ -61,11 +69,25 @@ public static Optional<OIDCInitialResponse> parse(String initialResponse) {
...
+            String userPart = null;
+            String associatedUserPart = null;
+
+            while (stringTokenizer.hasMoreTokens()) {
+                String stringToken = stringTokenizer.nextToken();
+                if (stringToken.startsWith("auth=")) {
+                    tokenPart = stringToken.substring(5).replace(PREFIX_TOKEN, "");
+                } else if (stringToken.startsWith("user=")) {
+                    userPart = stringToken.substring(5);
+                } else if (stringToken.startsWith("authzid=")) {
```
Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java:84]: Shouldn't this be `a=` ?

File: protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java [hunk 4]
```diff
@@ -61,11 +69,25 @@ public static Optional<OIDCInitialResponse> parse(String initialResponse) {
...
+
+            StringTokenizer stringTokenizer = new StringTokenizer(decodeValueWithoutDanglingPart, String.valueOf(SASL_SEPARATOR));
+            String tokenPart = null;
+            String userPart = null;
+            String associatedUserPart = null;
+
+            while (stringTokenizer.hasMoreTokens()) {
+                String stringToken = stringTokenizer.nextToken();
+                if (stringToken.startsWith("auth=")) {
+                    tokenPart = stringToken.substring(5).replace(PREFIX_TOKEN, "");
```
Comment (chibenwa) [inline protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java:81]: Arg we should not use `String::replace` as thisis backed by regexps.

Have a look at StringUtils from commons-lang 3 library

**Review by chibenwa — COMMENTED** [review]
Sorry I likely messed up the use of delegation over OIDC.

Me: extracted from the auth token claim (email)
Other user Id: `user=`

No `authidz`

File: protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java (hunk 3 above)
Comment (vttranlina) [reply protocols/api/src/main/java/org/apache/james/protocols/api/OIDCSASLParser.java]: `authidz=` -> `user=`

File: protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java [hunk 5]
```diff
@@ -98,26 +98,10 @@ void parseShouldReturnResponseWhenPartsDoNotSort() {
         SoftAssertions.assertSoftly(softly -> {
             softly.assertThat(parseResult.isPresent()).isTrue();
             softly.assertThat(parseResult.get().getToken()).isEqualTo(TOKEN_WITHOUT_PREFIX);
-            softly.assertThat(parseResult.get().getUser()).isEqualTo(USER);
+            softly.assertThat(parseResult.get().getAssociatedUser()).isEqualTo(USER);
         });
     }
 
-    @Test
-    void parseShouldReturnEmptyWhenInputHasSeveralAuthPart() {
```
Comment (chibenwa) [inline protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java:106]: Why is this test removed?

File: protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java [hunk 6]
```diff
@@ -98,26 +98,10 @@ void parseShouldReturnResponseWhenPartsDoNotSort() {
...
-        String input = buildOAuthBearerPlain(ImmutableList.of("n,auth=" + BEARER_TOKEN,
-            "user=" + USER,
-            "auth=token2"));
-        assertThat(OIDCSASLParser.parse(input)).isEmpty();
-    }
-
-    @Test
-    void parseShouldReturnEmptyWhenInputHasSeveralUserPart() {
```
Comment (chibenwa) [inline protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java:114]: Why is this test removed?

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java [hunk 7]
```diff
@@ -1211,6 +1219,72 @@ void oauthShouldFailWhenIntrospectTokenServerError() throws Exception {
...
+                String oauthBearer = OIDCSASLHelper.generateOauthBearer(USER2.asString(), OidcTokenFixture.VALID_TOKEN);
+                String authenticateResponse = client.sendCommand("AUTHENTICATE OAUTHBEARER " + oauthBearer);
+                assertThat(authenticateResponse).contains("OK AUTHENTICATE completed.");
+
+                assertThat(client.select("INBOX")
+                    .readFirstMessage()).contains(SMALL_MESSAGE);
+            }
+        }
+
+        private void createImapServerForImpersonateTest() throws Exception {
```
Comment (chibenwa) [inline server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1262]: Suggestion : always set this up in @BeforeEach.

For tests failing delegation you can use USER3 which USER have no access to.

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java [hunk 8]
```diff
@@ -1211,6 +1219,72 @@ void oauthShouldFailWhenIntrospectTokenServerError() throws Exception {
...
+        }
+
+        @Test
+        void oauthShouldImpersonateSuccessWhenDelegated() throws Exception {
+            createImapServerForImpersonateTest();
+
+            String oauthBearer = OIDCSASLHelper.generateOauthBearer(USER2.asString(), OidcTokenFixture.VALID_TOKEN);
+            IMAPSClient client = imapsClient(port);
+            client.sendCommand("AUTHENTICATE OAUTHBEARER " + oauthBearer);
+            assertThat(client.getReplyString()).contains("OK AUTHENTICATE completed.");
```
Comment (chibenwa) [inline server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1238]: Can we create a specific mailbox in user 2 account ? Then listing it after auth would ensure that we are well connected as USER2 and not as USER.

**Review by chibenwa — COMMENTED** [review]
Looks pretty good to me!

File: protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java (hunk 5 above)
Comment (vttranlina) [reply protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java:106]: When I use `StringTokenizer` for parsing parts, I see it is unnecessary,

File: server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java (hunk 8 above)
Comment (vttranlina) [reply server/protocols/protocols-imap4/src/test/java/org/apache/james/imapserver/netty/IMAPServerTest.java:1238]: It is already at `impersonationShouldWorkWhenDelegated`

**Comment by chibenwa** [comment]
Can we re add the two removed tests?

Then that is good for me.

File: protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java [hunk 9]
```diff
@@ -102,6 +101,21 @@ void parseShouldReturnResponseWhenPartsDoNotSort() {
...
+    }
+
+    @Test
+    void parseShouldReturnEmptyWhenInputHasSeveralUserPart() {
+        String input = buildOAuthBearerPlain(ImmutableList.of("n,user=" + USER,
+            "user=" + USER,
+            "auth=token2"));
+        assertThat(OIDCSASLParser.parse(input)).isEmpty();
+    }
     @ParameterizedTest
```
Comment (chibenwa) [inline protocols/api/src/test/java/org/apache/james/protocols/api/OIDCSASLParserTest.java:119]:
```suggestion
    }
    
    @ParameterizedTest
```

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]
Rebase please :)

**Review by quantranhong1999 — APPROVED** [review]

**Comment by chibenwa** [comment]
```
org.apache.james.protocols.lib.handler.ProtocolHandlerLoader$LoadingException: Unable to load protocolhandler
	at org.apache.james.smtpserver.DSNTest.orcptCanBeOmitted(DSNTest.java:294)
Caused by: com.google.inject.ConfigurationException: 
Guice configuration errors:

1) [Guice/MissingImplementation]: No implementation for Authorizator was bound.

Requested by:
1  : UsersRepositoryAuthHook.<init>(UsersRepositoryAuthHook.java:56)
      \_ for 2nd parameter
     while locating UsersRepositoryAuthHook

Learn more:
  https://github.com/google/guice/wiki/MISSING_IMPLEMENTATION

1 error

======================
Full classname legend:
======================
Authorizator:            "org.apache.james.mailbox.Authorizator"
UsersRepositoryAuthHook: "org.apache.james.smtpserver.UsersRepositoryAuthHook"
========================
End of classname legend:
```

Related...

**Comment by chibenwa** [comment]
`org.apache.james.mariadb.JPAMariaDBJamesServerTest.connectSMTPServerShouldSendShabangOnConnect`

```
org.apache.james.protocols.lib.handler.ProtocolHandlerLoader$LoadingException: Can not load org.apache.james.smtpserver.UsersRepositoryAuthHook
	at org.apache.james.mariadb.JPAMariaDBJamesServerTest.setup(JPAMariaDBJamesServerTest.java:57)
Caused by: com.google.inject.ConfigurationException: 
Guice configuration errors:

1) [Guice/MissingImplementation]: No implementation for Authorizator was bound.

Requested by:
1  : UsersRepositoryAuthHook.<init>(UsersRepositoryAuthHook.java:56)
      \_ for 2nd parameter
     while locating UsersRepositoryAuthHook
```

`org.apache.james.JPAJamesServerTest.connectSMTPServerShouldSendShabangOnConnect`

```
org.apache.james.protocols.lib.handler.ProtocolHandlerLoader$LoadingException: Can not load org.apache.james.smtpserver.UsersRepositoryAuthHook
	at org.apache.james.JPAJamesServerTest.setup(JPAJamesServerTest.java:77)
Caused by: com.google.inject.ConfigurationException: 
Guice configuration errors:

1) [Guice/MissingImplementation]: No implementation for Authorizator was bound.

Requested by:
1  : UsersRepositoryAuthHook.<init>(UsersRepositoryAuthHook.java:56)
      \_ for 2nd parameter
     while locating UsersRepositoryAuthHook

Learn more:
  https://github.com/google/guice/wiki/MISSING_IMPLEMENTATION

1 error

======================
Full classname legend:
======================
Authorizator:            "org.apache.james.mailbox.Authorizator"
UsersRepositoryAuthHook: "org.apache.james.smtpserver.UsersRepositoryAuthHook"
========================
End of classname legend:
========================
```

Can you have a look?

File: server/container/guice/jpa-common/src/main/java/org/apache/james/modules/data/JPANoAuthorizatorModule.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import com.google.inject.Provides;
+import com.google.inject.Singleton;
+
+public class JPANoAuthorizatorModule extends AbstractModule {
+
+    @Provides
+    @Singleton
+    Authorizator provideAuthorizator() {
+        return (userId, otherUserId) -> Authorizator.AuthorizationState.FORBIDDEN;
+    }
```
Comment (chibenwa) [inline server/container/guice/jpa-common/src/main/java/org/apache/james/modules/data/JPANoAuthorizatorModule.java:34]: UserRepositoryAuthorizator ?

**Review by Arsnael — APPROVED** [review]

**Pull request merged**